File: android/app/src/main/java/app/ripple/mesh/core/MeshRouter.kt
package app.ripple.mesh.core

import java.security.PublicKey
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/** A bidirectional byte pipe to one neighbour. Implemented by the BLE layer. */
interface Link {
    val id: String
    /** Hex NodeId of the peer, once learned from its ANNOUNCE. */
    var peerHex: String?
    fun send(packetBytes: ByteArray)
    fun close()
}

data class Peer(
    val nodeId: NodeId,
    val publicKey: PublicKey,
    val publicKeyWire: ByteArray,
    val name: String,
    val lastSeen: Long,
    /** Best known hop distance (1 = direct neighbour). */
    val hops: Int,
    val avatar: String? = null,
)

data class InboundMessage(
    val messageId: ByteArray,
    val from: NodeId,
    val fromName: String?,
    val text: String,
    val isBroadcast: Boolean,
    val verified: Boolean,
    val timestamp: Long,
)

data class InboundVoiceMessage(
    val messageId: ByteArray,
    val from: NodeId,
    val fromName: String?,
    val voiceBytes: ByteArray,
    val durationMs: Int,
    val isBroadcast: Boolean,
    val verified: Boolean,
    val timestamp: Long,
)

/** Delivery report for an SOS beacon. `location` is null unless the sender opted in to GPS. */
data class SosBeacon(
    val messageId: ByteArray,
    val from: NodeId,
    val fromName: String?,
    val text: String,
    val location: SosLocation?,
    val verified: Boolean,
    val timestamp: Long,
    val voiceBytes: ByteArray? = null,
    val voiceDurationMs: Int? = null,
)

interface RouterListener {
    fun onMessage(message: InboundMessage)
    fun onVoiceMessage(message: InboundVoiceMessage) {}
    fun onAck(messageId: ByteArray, from: NodeId)
    fun onPeersChanged(peers: List<Peer>)
    fun onLinkIdentified(link: Link, peer: Peer) {}
    fun onSos(beacon: SosBeacon) {}
}

/**
 * Transport-agnostic implementation of PROTOCOL.md §4: flooding with a seen-cache,
 * TTL, store-and-forward, duplicate-link suppression, signature verification and
 * end-to-end decryption. Identical in behaviour to tools/protocol/mesh-sim.js.
 *
 * Thread-safe: every public entry point takes the router lock. Outbound sends
 * happen while holding the lock, so [Link.send] must be non-blocking (queue + worker).
 */
class MeshRouter(
    val identity: Identity,
    displayName: String,
    private val listener: RouterListener,
    private val clock: () -> Long = System::currentTimeMillis,
    private val maxSeen: Int = 5_000,
    private val maxRelay: Int = 500,
) {
    @Volatile var displayName: String = displayName
        private set

    private class RelayEntry(val packet: Packet, val expiresAt: Long) { val deliveredTo = HashSet<String>() }

    private val lock = ReentrantLock()
    private val log = EventLog.global
    private val links = LinkedHashMap<String, Link>()
    private val peers = LinkedHashMap<String, Peer>()
    private val seen = object : LinkedHashMap<String, Long>(256, 0.75f, false) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, Long>) = size > maxSeen
    }
    private val relayStore = object : LinkedHashMap<String, RelayEntry>(64, 0.75f, false) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, RelayEntry>) = size > maxRelay
    }

    // Phase 2 policy state.
    private var batteryProfile: Int = BatteryProfile.BALANCED
    private var inboundRateCfg: Pair<Int, Long>? = null
    private val inboundRate = HashMap<String, RateLimiter>()

    val selfId: NodeId get() = identity.nodeId
    /** Store-and-forward retention (72 h, PROTOCOL.md §6). */
    val relayRetentionMs: Long get() = Protocol.RELAY_TTL_MS
    fun batteryProfileCode(): Int = lock.withLock { batteryProfile }
    fun setBatteryProfile(code: Int): Unit = lock.withLock {
        require(BatteryProfile.isValid(code)) { "bad battery profile $code" }
        batteryProfile = code
    }

    /** Cap how many broadcast/SOS packets a single source may push through this node per window. */
    fun setInboundRateLimit(max: Int, windowMs: Long): Unit = lock.withLock {
        require(max > 0 && windowMs > 0)
        inboundRateCfg = max to windowMs
        inboundRate.clear()
    }

    fun peers(): List<Peer> = lock.withLock { peers.values.toList() }
    fun peer(nodeId: NodeId): Peer? = lock.withLock { peers[nodeId.hex] }
    fun linkCount(): Int = lock.withLock { links.size }
    fun linkSnapshots(): List<Pair<String, String?>> = lock.withLock { links.values.map { it.id to it.peerHex } }
    fun directNeighbourCount(): Int = lock.withLock { links.values.count { it.peerHex != null } }

    fun setDisplayName(name: String): Unit = lock.withLock {
        displayName = name
        links.values.forEach { it.send(PacketFactory.announce(identity, name).encode()) }
    }

    /** Restore persisted peers (so direct messages can be sent before they re-announce). */
    fun importPeers(saved: Collection<Peer>): Unit = lock.withLock {
        saved.forEach { if (!peers.containsKey(it.nodeId.hex)) peers[it.nodeId.hex] = it }
    }

    /** Restore persisted relay-store packets after a restart. */
    fun importRelayStore(packets: Collection<Packet>): Unit = lock.withLock {
        val now = clock()
        packets.forEach { p ->
            seen[p.messageIdHex] = now
            relayStore[p.messageIdHex] = RelayEntry(p, now + Protocol.RELAY_TTL_MS)
        }
    }

    fun relayStoreSnapshot(): List<Packet> = lock.withLock { relayStore.values.map { it.packet } }

    // ---- outbound -----------------------------------------------------------------

    fun sendBroadcast(text: String): ByteArray = originate(PacketFactory.broadcastText(identity, text))

    /** @throws IllegalStateException if the peer's public key is unknown. */
    fun sendDirect(destination: NodeId, text: String): ByteArray {
        val peer = peer(destination) ?: throw IllegalStateException("Unknown peer ${destination.display}")
        return originate(PacketFactory.directText(identity, destination, peer.publicKeyWire, text))
    }

    fun sendBroadcastVoice(voiceBytes: ByteArray, durationMs: Int): ByteArray =
        originate(PacketFactory.broadcastVoice(identity, voiceBytes, durationMs))

    fun sendDirectVoice(destination: NodeId, voiceBytes: ByteArray, durationMs: Int): ByteArray {
        val peer = peer(destination) ?: throw IllegalStateException("Unknown peer ${destination.display}")
        return originate(PacketFactory.directVoice(identity, destination, peer.publicKeyWire, voiceBytes, durationMs))
    }

    /** Broadcast an SOS beacon. `location` is only sent if the operator opted in to sharing GPS. */
    fun sendSos(
        text: String,
        location: SosLocation? = null,
        voiceBytes: ByteArray? = null,
        voiceDurationMs: Int = 0
    ): ByteArray = originate(PacketFactory.sos(identity, text, location, voiceBytes, voiceDurationMs))

    private fun originate(packet: Packet): ByteArray = lock.withLock {
        if (packet.type != PacketType.ACK) log.i("router", "sending ${packet.type} ${packet.messageIdHex.take(8)} on ${links.size} link(s)")
        markSeen(packet)
        store(packet, fromLink = null)
        broadcast(packet, except = null)
        packet.messageId
    }

    // ---- link lifecycle -----------------------------------------------------------

    fun onLinkReady(link: Link): Unit = lock.withLock {
        links[link.id] = link
        link.send(PacketFactory.announce(identity, displayName).encode())
    }

    fun onLinkClosed(link: Link) = lock.withLock { links.remove(link.id); Unit }

    private fun onLinkIdentified(link: Link, peerHex: String) {
        // Duplicate-link suppression: the node with the larger id closes the newer link.
        val duplicate = links.values.any { it !== link && it.peerHex == peerHex }
        if (duplicate && selfId.hex > peerHex) { log.d("router", "closing duplicate link ${link.id} to ${peerHex.takeLast(4)}"); link.close(); links.remove(link.id); return }
        log.i("router", "link ${link.id} is ${peers[peerHex]?.name ?: peerHex.takeLast(4)}")

        peers[peerHex]?.let { listener.onLinkIdentified(link, it) }

        // Store-and-forward replay of anything this peer might still need.
        val peerId = NodeId.fromHex(peerHex)
        val now = clock()
        var replayed = 0
        for (entry in relayStore.values) {
            if (entry.expiresAt < now || peerHex in entry.deliveredTo) continue
            val dest = entry.packet.destination
            val forPeer = dest.hex == peerId.hex
            val peerCanRelay = dest.isBroadcast && (entry.packet.type == PacketType.SOS || batteryProfile != BatteryProfile.POWER_SAVER)
            if (forPeer || peerCanRelay) {
                entry.deliveredTo.add(peerHex)
                link.send(entry.packet.encode())
                replayed++
            }
        }
        if (replayed > 0) log.i("router", "replayed $replayed stored packet(s) to ${peerHex.takeLast(4)}")
    }

    // ---- inbound ------------------------------------------------------------------

    fun onReceive(link: Link, bytes: ByteArray): Unit = lock.withLock {
        val p = try { Packet.decode(bytes) } catch (e: Exception) { log.w("router", "bad packet from ${link.id}: ${e.message}"); return }
        val now = clock()
        if (p.timestamp > now + Protocol.SEEN_TTL_MS) return
        if (seen.containsKey(p.messageIdHex)) return
        markSeen(p)
        if (!rateAllows(p)) return // per-source broadcast flood cap (when configured)

        if (p.type == PacketType.ANNOUNCE) { handleAnnounce(link, p, now); return }

        val forMe = p.destination.hex == selfId.hex
        val isBroadcast = p.destination.isBroadcast

        if (p.type == PacketType.SOS) {
            if (!isBroadcast) { relay(p, link); return }
            val peer = peers[p.source.hex]
            val verified = peer != null && Crypto.verify(peer.publicKey, p.encodeUnsigned(), p.signature)
            if (peer != null && !verified) return // forged beacon from a known peer
            val sos = try { SosCodec.decode(p.payload) } catch (_: Exception) { relay(p, link); return }
            listener.onSos(SosBeacon(p.messageId, p.source, peer?.name, sos.text, sos.location, verified, p.timestamp, sos.voiceBytes, sos.voiceDurationMs))
            relay(p, link) // beacons always flood onward
            return
        }

        if (forMe || isBroadcast) {
            val peer = peers[p.source.hex]
            val verified = peer != null && Crypto.verify(peer.publicKey, p.encodeUnsigned(), p.signature)
            when {
                peer != null && !verified -> { log.w("router", "dropped ${p.type} ${p.messageIdHex.take(8)}: bad signature for ${p.source.short}"); return }
                forMe && peer == null -> { log.d("router", "${p.type} ${p.messageIdHex.take(8)} for me from unknown ${p.source.short}; relaying"); relay(p, link); return }
            }
            when (p.type) {
                PacketType.MESSAGE -> {
                    val text = if (p.isEncrypted) {
                        try { String(Crypto.decrypt(identity.privateKey, p.messageId, p.source, p.destination, p.payload), Charsets.UTF_8) }
                        catch (_: Exception) { log.w("router", "could not decrypt ${p.messageIdHex.take(8)} from ${p.source.short}"); return }
                    } else String(p.payload, Charsets.UTF_8)
                    log.i("router", "${if (isBroadcast) "broadcast" else "direct"} ${p.messageIdHex.take(8)} from ${peer?.name ?: p.source.short} via ${link.id} (ttl ${p.ttl})")
                    listener.onMessage(InboundMessage(p.messageId, p.source, peer?.name, text, isBroadcast, verified, p.timestamp))
                    if (forMe) originate(PacketFactory.ack(identity, p.source, p.messageId))
                }
                PacketType.VOICE -> {
                    val rawVoicePayload = if (p.isEncrypted) {
                        try { Crypto.decrypt(identity.privateKey, p.messageId, p.source, p.destination, p.payload) }
                        catch (_: Exception) { log.w("router", "could not decrypt voice ${p.messageIdHex.take(8)} from ${p.source.short}"); return }
                    } else p.payload
                    val (durationMs, voiceBytes) = try { VoiceCodec.decode(rawVoicePayload) }
                    catch (_: Exception) { log.w("router", "malformed voice payload from ${p.source.short}"); return }
                    log.i("router", "${if (isBroadcast) "broadcast" else "direct"} voice ${p.messageIdHex.take(8)} (${durationMs}ms) from ${peer?.name ?: p.source.short} via ${link.id} (ttl ${p.ttl})")
                    listener.onVoiceMessage(InboundVoiceMessage(p.messageId, p.source, peer?.name, voiceBytes, durationMs, isBroadcast, verified, p.timestamp))
                    if (forMe) originate(PacketFactory.ack(identity, p.source, p.messageId))
                }
                PacketType.ACK -> if (forMe && p.payload.size == Protocol.MESSAGE_ID_SIZE) listener.onAck(p.payload, p.source)
                PacketType.ANNOUNCE -> {}
                PacketType.SOS -> {} // handled above
            }
        }
        if (!forMe) relay(p, link)
    }

    private fun handleAnnounce(link: Link, p: Packet, now: Long) {
        val ann = try { Announce.decode(p.payload) } catch (_: Exception) { return }
        if (NodeId.fromPublicKey(ann.publicKeyWire).hex != p.source.hex) return
        val publicKey = try { Crypto.publicKeyFromWire(ann.publicKeyWire) } catch (_: Exception) { return }
        if (!Crypto.verify(publicKey, p.encodeUnsigned(), p.signature)) return
        if (p.source.hex == selfId.hex) return

        val (extractedAvatar, cleanName) = AvatarHelper.extractAvatarAndName(ann.name)
        val hops = Protocol.MAX_TTL - p.ttl + 1
        val prev = peers[p.source.hex]
        if (prev == null) log.i("router", "new peer ${ann.name} (${p.source.short}) at $hops hop(s)")
        peers[p.source.hex] = Peer(p.source, publicKey, ann.publicKeyWire, cleanName, now, if (prev != null) minOf(prev.hops, hops) else hops, extractedAvatar)
        listener.onPeersChanged(peers.values.toList())

        if (link.peerHex == null && hops == 1) {
            link.peerHex = p.source.hex
            onLinkIdentified(link, p.source.hex)
        }
        relay(p, link)
    }

    // ---- internals ----------------------------------------------------------------

    /** ANNOUNCEs and SOS beacons always relay (safety/liveness); ordinary chat does not in POWER_SAVER. */
    private fun isCritical(p: Packet) = p.type == PacketType.SOS || p.type == PacketType.ANNOUNCE

    private fun rateAllows(p: Packet): Boolean {
        val cfg = inboundRateCfg ?: return true
        if (p.type != PacketType.MESSAGE && p.type != PacketType.SOS) return true
        if (!p.destination.isBroadcast) return true // direct E2E is never flood-gated
        val limiter = inboundRate.getOrPut(p.source.hex) { RateLimiter(cfg.first, cfg.second) }
        return limiter.allow()
    }

    private fun relay(p: Packet, from: Link) {
        if (p.ttl <= 1) return
        if (batteryProfile == BatteryProfile.POWER_SAVER && !isCritical(p)) return
        val relayed = p.withTtl(p.ttl - 1)
        store(relayed, from)
        broadcast(relayed, from)
    }

    private fun broadcast(p: Packet, except: Link?) {
        val bytes = p.encode()
        val entry = relayStore[p.messageIdHex]
        for (link in links.values) {
            if (link === except) continue
            link.peerHex?.let { entry?.deliveredTo?.add(it) }
            link.send(bytes)
        }
    }

    private fun store(p: Packet, fromLink: Link?) {
        if (p.type == PacketType.ANNOUNCE) return
        val entry = RelayEntry(p, clock() + Protocol.RELAY_TTL_MS)
        fromLink?.peerHex?.let { entry.deliveredTo.add(it) }
        relayStore[p.messageIdHex] = entry
    }

    private fun markSeen(p: Packet) { seen[p.messageIdHex] = clock() }
}

File: ios/Ripple/Core/MeshRouter.swift
import Foundation
import CryptoKit

/// A bidirectional byte pipe to one neighbour. Implemented by the BLE layer.
protocol Link: AnyObject {
    var id: String { get }
    /// Hex NodeId of the peer once learned from its ANNOUNCE.
    var peerHex: String? { get set }
    /// Must be non-blocking (queue + worker).
    func send(_ packetBytes: Data)
    func close()
}

struct Peer {
    let nodeId: NodeId
    let publicKey: P256.Signing.PublicKey
    let publicKeyWire: Data
    let name: String
    let lastSeen: Date
    /// Best known hop distance (1 = direct neighbour).
    let hops: Int
}

struct InboundMessage {
    let messageId: Data
    let from: NodeId
    let fromName: String?
    let text: String
    let isBroadcast: Bool
    let verified: Bool
    let timestamp: UInt64
}

struct InboundVoiceMessage {
    let messageId: Data
    let from: NodeId
    let fromName: String?
    let voiceBytes: Data
    let durationMs: Int
    let isBroadcast: Bool
    let verified: Bool
    let timestamp: UInt64
}

/// Delivery report for an SOS beacon. `location` is nil unless the sender opted in to GPS.
struct SosBeacon {
    let messageId: Data
    let from: NodeId
    let fromName: String?
    let text: String
    let location: SosLocation?
    let verified: Bool
    let timestamp: UInt64
    let voiceBytes: Data?
    let voiceDurationMs: Int?

    init(messageId: Data, from: NodeId, fromName: String?, text: String, location: SosLocation?, verified: Bool, timestamp: UInt64, voiceBytes: Data? = nil, voiceDurationMs: Int? = nil) {
        self.messageId = messageId
        self.from = from
        self.fromName = fromName
        self.text = text
        self.location = location
        self.verified = verified
        self.timestamp = timestamp
        self.voiceBytes = voiceBytes
        self.voiceDurationMs = voiceDurationMs
    }
}

protocol RouterListener: AnyObject {
    func router(_ router: MeshRouter, didReceive message: InboundMessage)
    func router(_ router: MeshRouter, didReceiveVoice message: InboundVoiceMessage)
    func router(_ router: MeshRouter, didReceiveAck messageId: Data, from: NodeId)
    func router(_ router: MeshRouter, peersDidChange peers: [Peer])
    func router(_ router: MeshRouter, didIdentify link: Link, as peer: Peer)
    func router(_ router: MeshRouter, didReceiveSos beacon: SosBeacon)
}

extension RouterListener {
    func router(_ router: MeshRouter, didReceiveVoice message: InboundVoiceMessage) {}
    func router(_ router: MeshRouter, didIdentify link: Link, as peer: Peer) {}
    func router(_ router: MeshRouter, didReceiveSos beacon: SosBeacon) {}
}

/// Transport-agnostic implementation of PROTOCOL.md §4. Behaviourally identical to
/// the Kotlin `MeshRouter` and `tools/protocol/mesh-sim.js`.
///
/// Thread-safe: every public entry point runs on the router's serial queue.
final class MeshRouter {
    let identity: Identity
    private(set) var displayName: String
    weak var listener: RouterListener?
    private let now: () -> Date
    private let maxSeen: Int
    private let maxRelay: Int

    private final class RelayEntry {
        let packet: Packet; let expiresAt: Date; var deliveredTo = Set<String>()
        init(_ p: Packet, expiresAt: Date) { packet = p; self.expiresAt = expiresAt }
    }

    private let queue = DispatchQueue(label: "app.ripple.mesh.router")
    private let log = EventLog.global
    private static let queueKey = DispatchSpecificKey<Void>()

    /// `queue.sync` that tolerates re-entrancy (e.g. a link closed from inside the
    /// router calling back into `onLinkClosed`). Mirrors Kotlin's ReentrantLock.
    private func locked<T>(_ body: () throws -> T) rethrows -> T {
        if DispatchQueue.getSpecific(key: Self.queueKey) != nil { return try body() }
        return try queue.sync(execute: body)
    }
    private var links: [String: Link] = [:]
    private var linkOrder: [String] = []
    private var peers: [String: Peer] = [:]
    private var seen: [String: Date] = [:]
    private var seenOrder: [String] = []
    private var relayStore: [String: RelayEntry] = [:]
    private var relayOrder: [String] = []

    // Phase 2 policy state.
    private var batteryProfile = BatteryProfile.balanced
    private var inboundRateCfg: (max: Int, windowMs: TimeInterval)?
    private var inboundRate: [String: RateLimiter] = [:]

    /// Store-and-forward retention (72 h, PROTOCOL.md §6).
    var relayRetention: TimeInterval { MeshProtocol.relayTTL }
    var batteryProfileCode: Int { batteryProfile.rawValue }
    func setBatteryProfile(_ code: Int) {
        locked {
            guard let p = BatteryProfile(rawValue: code) else { return }
            batteryProfile = p
        }
    }

    /// Cap how many broadcast/SOS packets a single source may push through this node per window.
    func setInboundRateLimit(max: Int, windowMs: TimeInterval) {
        locked {
            inboundRateCfg = (max, windowMs)
            inboundRate.removeAll()
        }
    }

    init(identity: Identity, displayName: String, listener: RouterListener?, now: @escaping () -> Date = Date.init, maxSeen: Int = 5000, maxRelay: Int = 500) {
        self.identity = identity; self.displayName = displayName; self.listener = listener
        self.now = now; self.maxSeen = maxSeen; self.maxRelay = maxRelay
        queue.setSpecific(key: Self.queueKey, value: ())
    }

    var selfId: NodeId { identity.nodeId }
    func allPeers() -> [Peer] { locked { Array(peers.values) } }
    func peer(_ id: NodeId) -> Peer? { locked { peers[id.hex] } }
    func linkCount() -> Int { locked { links.count } }
    func linkSnapshots() -> [(id: String, peerHex: String?)] { locked { linkOrder.compactMap { id in links[id].map { ($0.id, $0.peerHex) } } } }
    func directNeighbourCount() -> Int { locked { links.values.filter { $0.peerHex != nil }.count } }

    func setDisplayName(_ name: String) {
        locked {
            displayName = name
            guard let ann = try? PacketFactory.announce(identity, name: name).encode() else { return }
            links.values.forEach { $0.send(ann) }
        }
    }

    func importPeers(_ saved: [Peer]) {
        locked { for p in saved where peers[p.nodeId.hex] == nil { peers[p.nodeId.hex] = p } }
    }

    func importRelayStore(_ packets: [Packet]) {
        locked {
            let t = now()
            for p in packets {
                markSeen(p, at: t)
                putRelay(p.messageIdHex, RelayEntry(p, expiresAt: t.addingTimeInterval(MeshProtocol.relayTTL)))
            }
        }
    }

    func relayStoreSnapshot() -> [Packet] { locked { relayOrder.compactMap { relayStore[$0]?.packet } } }

    // MARK: Outbound

    @discardableResult
    func sendBroadcast(_ text: String) throws -> Data {
        originate(try PacketFactory.broadcastText(identity, text))
    }

    /// Broadcast an SOS beacon. `location` is only sent if the operator opted in to sharing GPS.
    @discardableResult
    func sendSos(text: String, location: SosLocation? = nil, voiceBytes: Data? = nil, voiceDurationMs: Int = 0) throws -> Data {
        originate(try PacketFactory.sos(identity, text: text, location: location, voiceBytes: voiceBytes, voiceDurationMs: voiceDurationMs))
    }

    enum SendError: Error { case unknownPeer(NodeId) }

    @discardableResult
    func sendDirect(to destination: NodeId, text: String) throws -> Data {
        guard let peer = peer(destination) else { throw SendError.unknownPeer(destination) }
        return originate(try PacketFactory.directText(identity, to: destination, recipientWire: peer.publicKeyWire, text: text))
    }

    @discardableResult
    func sendBroadcastVoice(durationMs: Int, opusBytes: Data) throws -> Data {
        originate(try PacketFactory.broadcastVoice(identity, durationMs: durationMs, opusBytes: opusBytes))
    }

    @discardableResult
    func sendDirectVoice(to destination: NodeId, durationMs: Int, opusBytes: Data) throws -> Data {
        guard let peer = peer(destination) else { throw SendError.unknownPeer(destination) }
        return originate(try PacketFactory.directVoice(identity, to: destination, recipientWire: peer.publicKeyWire, durationMs: durationMs, opusBytes: opusBytes))
    }

    @discardableResult
    func sendAck(to destination: NodeId, acknowledged messageId: Data) throws -> Data {
        originate(try PacketFactory.ack(identity, to: destination, acknowledged: messageId))
    }

    private func originate(_ packet: Packet) -> Data {
        locked { originateLocked(packet) }
    }

    private func originateLocked(_ packet: Packet) -> Data {
        if packet.type != .ack { log.i("router", "sending \(packet.type) \(packet.messageIdHex.prefix(8)) on \(links.count) link(s)") }
        markSeen(packet, at: now())
        store(packet, from: nil)
        broadcast(packet, except: nil)
        return packet.messageId
    }

    // MARK: Link lifecycle

    func onLinkReady(_ link: Link) {
        locked {
            links[link.id] = link; linkOrder.append(link.id)
            if let ann = try? PacketFactory.announce(identity, name: displayName).encode() { link.send(ann) }
        }
    }

    func onLinkClosed(_ link: Link) {
        locked { removeLink(link.id) }
    }

    private func removeLink(_ id: String) {
        links[id] = nil; linkOrder.removeAll { $0 == id }
    }

    private func onLinkIdentified(_ link: Link, peerHex: String) {
        let duplicate = links.values.contains { $0 !== link && $0.peerHex == peerHex }
        if duplicate && selfId.hex > peerHex { log.d("router", "closing duplicate link \(link.id) to \(peerHex.suffix(4))"); link.close(); removeLink(link.id); return }
        log.i("router", "link \(link.id) is \(peers[peerHex]?.name ?? String(peerHex.suffix(4)))")

        if let p = peers[peerHex] { listener?.router(self, didIdentify: link, as: p) }

        let t = now()
        var replayed = 0
        defer { if replayed > 0 { log.i("router", "replayed \(replayed) stored packet(s) to \(peerHex.suffix(4))") } }
        for key in relayOrder {
            guard let entry = relayStore[key], entry.expiresAt > t, !entry.deliveredTo.contains(peerHex) else { continue }
            let dest = entry.packet.destination
            let forPeer = dest.hex == peerHex
            let unknownDest = !dest.isBroadcast && peers[dest.hex] == nil
            if forPeer || dest.isBroadcast || unknownDest {
                entry.deliveredTo.insert(peerHex)
                link.send(entry.packet.encode())
                replayed += 1
            }
        }
    }

    // MARK: Inbound

    func onReceive(_ link: Link, _ bytes: Data) {
        locked { receiveLocked(link, bytes) }
    }

    private func receiveLocked(_ link: Link, _ bytes: Data) {
        guard let p = try? Packet.decode(bytes) else { return }
        let t = now()
        if Double(p.timestamp) / 1000 > t.timeIntervalSince1970 + MeshProtocol.seenTTL { return }
        if seen[p.messageIdHex] != nil { return }
        markSeen(p, at: t)
        guard rateAllows(p) else { return } // per-source broadcast flood cap (when configured)

        if p.type == .announce { handleAnnounce(link, p, at: t); return }

        let forMe = p.destination == selfId
        let isBroadcast = p.destination.isBroadcast

        if p.type == .sos {
            if !isBroadcast { relay(p, from: link); return }
            let peer = peers[p.source.hex]
            let verified = peer.map { Crypto.verify($0.publicKey, unsignedPacket: p.encodeUnsigned(), rawSignature: p.signature) } ?? false
            if peer != nil && !verified { return } // forged beacon from a known peer
            guard let sos = try? SosCodec.decode(p.payload) else { relay(p, from: link); return }
            listener?.router(self, didReceiveSos: SosBeacon(messageId: p.messageId, from: p.source, fromName: peer?.name,
                                                            text: sos.text, location: sos.location, verified: verified, timestamp: p.timestamp,
                                                            voiceBytes: sos.voiceBytes, voiceDurationMs: sos.voiceDurationMs))
            relay(p, from: link) // beacons always flood onward
            return
        }

        if forMe || isBroadcast {
            let peer = peers[p.source.hex]
            let verified = peer.map { Crypto.verify($0.publicKey, unsignedPacket: p.encodeUnsigned(), rawSignature: p.signature) } ?? false
            if peer != nil && !verified { log.w("router", "dropped \(p.type) \(p.messageIdHex.prefix(8)): bad signature for \(p.source.short)"); return }
            if forMe && peer == nil { log.d("router", "\(p.type) \(p.messageIdHex.prefix(8)) for me from unknown \(p.source.short); relaying"); relay(p, from: link); return }

            switch p.type {
            case .message:
                let text: String
                if p.isEncrypted {
                    guard let plain = try? Crypto.decrypt(recipient: identity.agreement, messageId: p.messageId, source: p.source, destination: p.destination, payload: p.payload),
                          let s = String(data: plain, encoding: .utf8) else { log.w("router", "could not decrypt \(p.messageIdHex.prefix(8)) from \(p.source.short)"); return }
                    text = s
                } else {
                    guard let s = String(data: p.payload, encoding: .utf8) else { return }
                    text = s
                }
                log.i("router", "\(isBroadcast ? "broadcast" : "direct") \(p.messageIdHex.prefix(8)) from \(peer?.name ?? p.source.short) via \(link.id) (ttl \(p.ttl))")
                listener?.router(self, didReceive: InboundMessage(messageId: p.messageId, from: p.source, fromName: peer?.name, text: text, isBroadcast: isBroadcast, verified: verified, timestamp: p.timestamp))
                if forMe, let ack = try? PacketFactory.ack(identity, to: p.source, acknowledged: p.messageId) { _ = originateLocked(ack) }
            case .voice:
                let voiceData: Data
                if p.isEncrypted {
                    guard let plain = try? Crypto.decrypt(recipient: identity.agreement, messageId: p.messageId, source: p.source, destination: p.destination, payload: p.payload) else {
                        log.w("router", "could not decrypt voice \(p.messageIdHex.prefix(8)) from \(p.source.short)")
                        return
                    }
                    voiceData = plain
                } else {
                    voiceData = p.payload
                }
                guard let decoded = try? VoiceCodec.decode(voiceData) else {
                    log.w("router", "malformed voice packet \(p.messageIdHex.prefix(8))")
                    return
                }
                log.i("router", "\(isBroadcast ? "broadcast" : "direct") voice \(p.messageIdHex.prefix(8)) from \(peer?.name ?? p.source.short) via \(link.id) (ttl \(p.ttl))")
                listener?.router(self, didReceiveVoice: InboundVoiceMessage(messageId: p.messageId, from: p.source, fromName: peer?.name, voiceBytes: decoded.opusBytes, durationMs: decoded.durationMs, isBroadcast: isBroadcast, verified: verified, timestamp: p.timestamp))
                if forMe, let ack = try? PacketFactory.ack(identity, to: p.source, acknowledged: p.messageId) { _ = originateLocked(ack) }
            case .ack:
                if forMe && p.payload.count == MeshProtocol.messageIdSize { listener?.router(self, didReceiveAck: p.payload, from: p.source) }
            case .announce:
                break
            case .sos:
                break // handled above
            }
        }
        if !forMe { relay(p, from: link) }
    }

    private func handleAnnounce(_ link: Link, _ p: Packet, at t: Date) {
        guard let ann = try? Announce.decode(p.payload),
              NodeId.fromPublicKey(ann.publicKeyWire) == p.source,
              let key = try? Crypto.signingKey(fromWire: ann.publicKeyWire),
              Crypto.verify(key, unsignedPacket: p.encodeUnsigned(), rawSignature: p.signature),
              p.source != selfId else { return }

        let hops = Int(MeshProtocol.maxTTL) - Int(p.ttl) + 1
        let prev = peers[p.source.hex]
        if prev == nil { log.i("router", "new peer \(ann.name) (\(p.source.short)) at \(hops) hop(s)") }
        peers[p.source.hex] = Peer(nodeId: p.source, publicKey: key, publicKeyWire: ann.publicKeyWire, name: ann.name, lastSeen: t, hops: prev.map { min($0.hops, hops) } ?? hops)
        listener?.router(self, peersDidChange: Array(peers.values))

        if link.peerHex == nil && hops == 1 {
            link.peerHex = p.source.hex
            onLinkIdentified(link, peerHex: p.source.hex)
        }
        relay(p, from: link)
    }

    // MARK: Internals

    /// ANNOUNCEs and SOS beacons always relay (safety/liveness); ordinary chat does not in POWER_SAVER.
    private func isCritical(_ p: Packet) -> Bool {
        p.type == .sos || p.type == .announce
    }

    private func rateAllows(_ p: Packet) -> Bool {
        guard let cfg = inboundRateCfg else { return true }
        if p.type != .message && p.type != .sos && p.type != .voice { return true }
        if !p.destination.isBroadcast { return true } // direct E2E is never flood-gated
        let limiter: RateLimiter
        if let existing = inboundRate[p.source.hex] {
            limiter = existing
        } else {
            let l = RateLimiter(max: cfg.max, windowMs: cfg.windowMs)
            inboundRate[p.source.hex] = l
            limiter = l
        }
        return limiter.allow()
    }

    private func relay(_ p: Packet, from: Link) {
        guard p.ttl > 1 else { return }
        if batteryProfile == .powerSaver && !isCritical(p) { return }
        let relayed = p.withTTL(p.ttl - 1)
        store(relayed, from: from)
        broadcast(relayed, except: from)
    }

    private func broadcast(_ p: Packet, except: Link?) {
        let bytes = p.encode()
        let entry = relayStore[p.messageIdHex]
        for id in linkOrder {
            guard let link = links[id], link !== except else { continue }
            if let ph = link.peerHex { entry?.deliveredTo.insert(ph) }
            link.send(bytes)
        }
    }

    private func store(_ p: Packet, from: Link?) {
        guard p.type != .announce else { return }
        let entry = RelayEntry(p, expiresAt: now().addingTimeInterval(MeshProtocol.relayTTL))
        if let ph = from?.peerHex { entry.deliveredTo.insert(ph) }
        putRelay(p.messageIdHex, entry)
    }

    private func putRelay(_ key: String, _ entry: RelayEntry) {
        if relayStore[key] == nil { relayOrder.append(key) }
        relayStore[key] = entry
        while relayOrder.count > maxRelay { relayStore[relayOrder.removeFirst()] = nil }
    }

    private func markSeen(_ p: Packet, at t: Date) {
        if seen[p.messageIdHex] == nil { seenOrder.append(p.messageIdHex) }
        seen[p.messageIdHex] = t
        while seenOrder.count > maxSeen { seen[seenOrder.removeFirst()] = nil }
    }
}
